File: AEEDs2/TP1/PalindromoRecursivo.java
/*
 * Nome: Victor Ferraz de Moraes
 * Matrícula: 802371
 * Curso: Ciência da Computação
 */

import java.util.Scanner;

public class PalindromoRecursivo {
    // Função recursiva que verifica se uma string é um palíndromo
    public static boolean isPalindromo(String palavra) {
        // Caso base: uma string vazia ou com um único caractere é sempre um palíndromo
        if (palavra.length() <= 1) {
            return true;
        } else {
            // Compara o primeiro e o último caractere da string
            char primeiro = palavra.charAt(0);
            char ultimo = palavra.charAt(palavra.length() - 1);
            
            // Se os caracteres forem diferentes, a string não é um palíndromo
            if (primeiro != ultimo) {
                return false;
            } else {
                // Chama a função recursivamente com a substring sem o primeiro e último caractere
                String subPalavra = palavra.substring(1, palavra.length() - 1);
                return isPalindromo(subPalavra);
            }
        }
    }

    public static void main(String args[]) {
        Scanner sc = new Scanner(System.in);
        String entrada = sc.nextLine();
        // Lê strings da entrada padrão até encontrar a palavra "FIM"
        while (!entrada.equals("FIM")) {
            if (isPalindromo(entrada)) {
                System.out.println("SIM"); // Se for um palíndromo, imprime "SIM"
            } else {
                System.out.println("NAO"); // Se não for um palíndromo, imprime "NAO"
            }
            entrada = sc.nextLine(); // Lê a próxima entrada
        }
        sc.close(); // Fecha o scanner quando não precisamos mais
    }
}

File: AEEDs2/TP1/PalindromoRecursivo.c
/*
 * Nome: Victor Ferraz de Moraes
 * Matrícula: 802371
 * Curso: Ciência da Computação
 */

#include <stdio.h>
#include <stdbool.h>
#include <string.h>

// Função recursiva para verificar se uma string é um palíndromo
bool isPalindromo(char palavra[], int inicio, int fim) {
    // Caso base: quando o índice de início é maior ou igual ao índice de fim, a string é um palíndromo
    if (inicio >= fim) {
        return true;
    }

    // Verifica se os caracteres nas posições inicio e fim são iguais
    if (palavra[inicio] != palavra[fim]) {
        return false; // Se forem diferentes, a string não é um palíndromo
    }

    // Chamada recursiva para verificar os caracteres internos da string
    return isPalindromo(palavra, inicio + 1, fim - 1);
}

int main() {
    char palavra[900]; // Declara um array de caracteres para armazenar a palavra
    scanf("%[^\r\n]%*c", palavra); // Lê a primeira palavra da entrada (ignorando o caractere de nova linha)
    while (strcmp(palavra, "FIM") != 0) { // Enquanto a palavra lida não for "FIM"
        int tam = strlen(palavra);
        bool resultado = isPalindromo(palavra, 0, tam - 1); // Chama a função recursiva

        if (resultado) {
            printf("SIM\n"); // Se for um palíndromo, imprime "SIM"
        } else {
            printf("NAO\n"); // Se não for um palíndromo, imprime "NAO"
        }
        scanf("%[^\r\n]%*c", palavra); // Lê a próxima palavra da entrada (ignorando o caractere de nova linha)
    }

    return 0;
}
